src/Sylius/Bundle/UserBundle/Form/EventListener/GuestCustomerFormSubscriber.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\UserBundle\Form\EventListener;
+
+use Sylius\Component\Resource\Factory\FactoryInterface;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+use Sylius\Component\User\Context\CustomerContextInterface;
+use Sylius\Component\User\Model\CustomerInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormEvents;
+use Symfony\Component\Form\FormInterface;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+class GuestCustomerFormSubscriber implements EventSubscriberInterface
+{
+    /**
+     * @var RepositoryInterface
+     */
+    private $customerRepository;
+
+    /**
+     * @var FactoryInterface
+     */
+    private $customerFactory;
+
+    /**
+     * @var CustomerContextInterface
+     */
+    private $customerContext;
+
+    /**
+     * @param RepositoryInterface      $customerRepository
+     * @param FactoryInterface         $customerFactory
+     * @param CustomerContextInterface $customerContext
+     */
+    public function __construct(RepositoryInterface $customerRepository, FactoryInterface $customerFactory, CustomerContextInterface $customerContext)
+    {
+        $this->customerRepository = $customerRepository;
+        $this->customerFactory = $customerFactory;
+        $this->customerContext = $customerContext;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function getSubscribedEvents()
+    {
+        return array(
+            FormEvents::PRE_SUBMIT => 'preSubmit',
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function preSubmit(FormEvent $event)
+    {
+        $rawData = $event->getData();
+        $form = $event->getForm();
+
+        $customer = $this->getCustomerFromProperSource($rawData, $form);
+
+        $form->setData($customer);
+    }
+
+    /**
+     * @param array         $rawData
+     * @param FormInterface $form
+     *
+     * @return CustomerInterface|null
+     */
+    protected function getCustomerFromProperSource($rawData, FormInterface $form)
+    {
+        if (null !== $customer = $this->customerContext->getCustomer()) {
+            $form->remove('email');
+
+            return $customer;
+        }
+
+        if (!isset($rawData['email'][0])) {
```

Comment:
For what you are looking here? Seems wrong as below you don't use just first element but whole data.